<!DOCTYPE html>
<html>

<head>
    <meta charset="UTF-8">
    <link rel="stylesheet" type="text/css" href="css/bootstrap.min.css">
    <link rel="stylesheet" type="text/css" href="css/scroll-top.css">
    <link rel="stylesheet" type="text/css" href="css/normalize.css">
    <link rel="stylesheet" type="text/css" href="css/style.css">
    <script type="text/x-mathjax-config">
        MathJax.Hub.Config({ tex2jax: {inlineMath: [['$','$'], ['\\(','\\)']]} });
    </script>
    <script type="text/javascript" async src="node_modules/mathjax/MathJax.js?config=TeX-MML-AM_CHTML"></script>
    <script type="text/javascript" src="node_modules/clipboard/dist/clipboard.min.js"></script>
    <script type="text/javascript" src="js/jquery.min.js"></script>
    <script type="text/javascript" src="js/bootstrap.min.js"></script>
    <script type="text/javascript" src="js/citetag.js"></script>
    <script type="text/javascript" src="js/header-scroll.js"></script>
</head>

<body>

    <header id="header">
        <h1>2016 Lab Notebook</h1>
    </header>

    <summary id="contents">
        <h2>Contents</h2>
        <table id="toc" class="no-centre">
            <tr>
                <td><a href="#2016-10">October</a></td>
                <td></td>
            </tr>
            <tr>
                <td></td>
                <td><a href="#2016-10_project">Project</a></td>
            </tr>
            <tr>
                <td></td>
                <td><a href="#2016-10_setup">Setup</a></td>
            </tr>
            <tr>
                <td></td>
                <td><a href="#2016-10_c3d-results">C3D Results</a></td>
            </tr>
            <tr>
                <td></td>
                <td><a href="#2016-10_MuSE">MuSE</a></td>
            </tr>
            <tr>
                <td></td>
                <td><a href="#2016-10_c3d-recommendations">C3D Recommendations</a></td>
            </tr>
        </table>
    </summary>
    <hr>

    <article>
        <section id="2016-10">
            <h2>October</h2>

            <section id="2016-10_project">
                <h3>Project</h3>
                <p><abbr title="Deoxyribonucleic acid">DNA</abbr> is more than just a series of nucleotides; it's a physical object with physical properties. DNA, along with other molecules and amino acids forms a structure called chromatin.</p>
                <p>This physical structure means that DNA translation is not just affected by the DNA sequences, or the DNA sequences nearby. There can be regulatory elements of DNA far off, that come into close proximity of a gene of interest just dependent on how the chromatin coils.</p>
                <p>Sven has recently published a paper on using this idea in order to find an <abbr title="Set of Regulatory Elements">SRE</abbr> for a particular gene of interest. He used two main tools to do this, <abbr title="Cross-Cell-type Correlation in DNase hypersensitivity signals">C3D</abbr>, and <abbr title="mutational significance within the regulatory element set">MuSE</abbr>. C3D finds correlations between <abbr title="DNase Hypersensitive Sites">DHS</abbr> within a certain window of a gene of interest. This helps give some context as to what the 3D strucuture of the chromatin is, as the hypothesis is that this SRE will contain regions of chromatin in close physical proximity to the gene. MuSE looks at the <abbr title="Background Mutation Rate">BMR</abbr> both globally and locally, and attempts to determine whether mutations occuring in an SRE appear significantly more often than in different parts of the genome.</p>
                <p>Sven's paper used 98 <abbr title="Breast Cancer">BCa</abbr> samples, and I'm going to try and replicate this analysis on a ~560 sample BCa dataset from <abbr title="International Cancer Genome Consortium">ICGC</abbr>.</p>
            </section>
            
            <section id="2016-10_setup">
                <h3>Setup</h3>
                <p>Tahmid (a co-op student from Waterloo), showed me the layout of the project, pointed me towards the scripts, and where the data is located. <strong class="danger">The lab doesn't use version control (!!!!) </strong> so the scripts are just located in directories on Mordor. I've copied important items into this repo. (Note: <code>$LHOME=/mnt/work1/users/lupiengroup/hawleyj</code>)</p>
                <table style="width: 100%;">
                    <tr>
                        <th>Item</th>
                        <th>Location</th>
                    </tr>
                    <tr>
                        <td>Original C3D</td>
                        <td><code>/mnt/work1/users/lupiengroup/TahmidProject/C3D/</code></td>
                    </tr>
                    <tr>
                        <td>Original MuSE</td>
                        <td><code>/mnt/work1/users/lupiengroup/TahmidProject/MUSE/</code></td>
                    </tr>
                    <tr>
                        <td>Original Somatic Mutations (Sanger)</td>
                        <td><code>/mnt/work1/users/pughlab/ICGC/BRCA-EU-UK_ICGC_May-2016/</code></td>
                    </tr>
                    <tr>
                        <td>BCa DHS BigWigs</td>
                        <td><code>/mnt/work1/data/commondata/ENCODE/data_ENCODE_ucsc/hg19_encodeAnalysisHub_build37/UniformSignalsFor_DNase_FAIRE_Histone_TFBS/UniformSignals</code></td>
                    </tr>
                    <tr>
                        <td>BCa DHS BedGraphs (converted from BigWig)</td>
                        <td><code>/mnt/work1/users/lupiengroup/TahmidProject/UW_DHS_BG/</code></td>
                    </tr>
                    <tr>
                        <td>Genes List for Promoters</td>
                        <td><code>/mnt/work1/users/lupiengroup/TahmidProject/genes.list</code></td>
                    </tr>
                    <tr>
                        <td>MCF-7 Cell Lines</td>
                        <td><code>/mnt/work1/data/commondata/ENCODE/data_ENCODE_ucsc/hg19_encodeProductionData_build37/Chromatin_Accessibility/UW_DNaseI_HS</code></td>
                    </tr>
                    <caption>Table 1: Original data locations. Small enough files of import are in <code>$LHOME/Data</code>, and large samples were left in their original locations.</caption>
                </table>

                <p>I made <code>promoter_anchors.bed</code> from running the <code>make_anchors_from_genes.sh</code> script:</p>
                <pre><code>
sh make_anchors_from_genes.sh genes.list 2500 500 promoter_anchors.bed
sort -k1,1 -k2,2n promoter_anchors.bed > promoter_anchors.sorted.bed                # sorted
bedtools merge -i promoter_anchors.sorted.bed > promoter_anchors.merged.sorted.bed  # merge to reduce features
                </code></pre>

                <p>Then, after setting up the configuration file <a href="files/79_BCa_samples.cfg">79_BCa_samples.cfg</a>, I ran C3D:</p>
                <pre><code>
sh c3d_wrapper.sh 79_BCa_samples.cfg
                </code></pre>
            </section>

            <section id="2016-10_c3d-results">
                <h3>C3D Results</h3>

                <p>Its initial run completed and produced <abbr title="Browser Extensible Data">BED</abbr> files and <code>signalMatrix.txt</code>. Tahmid said that this allows for faster runs after the first pass. It failed though, running <code>c3d.R</code> becuase it couldn't find <code>JASPAR_IDs.txt</code>, which the R script assumes is in the same directory. I'm rerunning C3D via <code>sh c3d_wrapper.sh 79_BCa_samples_rec.cfg</code> to properly reference the JASPAR file.</p>

                <h4>Mistakes were made</h4>
                <p>Even if C3D hadn't crashed at the Rscript, it would've taken weeks to run, because I accidentally used the unmerged/unsorted anchor file. That has been corrected in the config files.</p>
                <p>I also ended up corrupting the output files when trying to rerun. I still have the original signalMatrix.txt and most of the reference BEDs, but I think a few of them were overwritten.</p>
                <p>To fix these mistakes, I've made 8 different folders, each containing a portion of the merged/sorted anchor file. It's a poor man's parallelization implementation, but given the current state of C3D I think it's appropriate. Parallelizing it would require an entire rewrite of the code with lots of abstraction. I think it's necessary, just not for me and not right now.</p>
                <p>It took 6.5h to map the BED files, and about 3h for the Rscript to parse 1000 anchors. Given that I'm restarting the mapping on all 8 sets, and each set contains 8236 anchors, it should take around 30h to complete. That translates to finishing at 22:00 on 2016-10-25.</p>
                <p>Turns out I forgot to modify the config files for sections 3-7. So those failed entering the Rscript. In order to avoid potentially corrupting the data again, I made copies of the directories, fixed the config files, then ran the scripts again. So they should finish sometime tomorrow (2016-10-26).</p>

                <p>This finally worked. I copied all the resulting files into <code>20161026_C3D</code>, and formatted the files for input to MuSE via (replace <code>$i</code> with each number):</p>
                <pre><code>
cat results_$i.txt | awk -v OFS='\t' '{if (NR!=1) {print $1, $4, $7, $7, $8}}' > results_$i.formatted.txt
cat results_*.formatted.txt >> results_all.txt
                </code></pre>
            </section>

            <section id="2016-10_MuSE">
                <h3>MuSE</h3>

                <p>Looking at the <code>simple_somatic_mutation.open.BRCA-EU.tsv.gz</code> file Sven pointed me towards, I'm using that, plus the C3D results from before to look at significant regions. I need to format the TSV for MuSE to read it properly. To do that, I ran:</p>
                <pre><code>
# collect the appropriate columns, prepend "chr" to chromosome column, remove headers
zcat simple_somatic_mutation.open.BRCA-EU.tsv.gz | awk -F "\t" 'BEGIN{OFS="\t";} {$9 = "chr" $9; if (NR != 1) {print $9, $10, $11, $7, $16, $17}}' > simple_somatic_mutation.open.BRCA-EU.bed
# sort BED file
sort -k1,1 -k2,2n simple_somatic_mutation.open.BRCA-EU.bed > simple_somatic_mutation.open.BRCA-EU.sorted.bed
                </code></pre>

                <p>I've then run MuSE as follows:</p>
                    <!--<button class="btn" data-clipboard-snippet=""><img class="clippy" width="13" src="assets/clippy.svg" alt="Copy to clipboard"></button>-->
                <pre><code>
qsub run-MuSE.sh /mnt/work1/users/lupiengroup/hawleyj/Data/simple_somatic_mutation.open.BRCA-EU.sorted.bed results_all.formatted.txt /mnt/work1/users/lupiengroup/hawleyj/Data/wgEncodeUwDnaseMcf7HotspotsRep1.broadPeak 500000 0.7
                </code></pre>

                <p>To filter out duplicate mutations, sort with the unique option:</p>
                <pre><code>
sort -u -k1,1 -k2,2n simple_somatic_mutation.open.BRCA-EU.bed > simple_somatic_mutation.open.BRCA-EU.sorted.unique.bed
qsub run-MuSE.sh /mnt/work1/users/lupiengroup/hawleyj/Data/simple_somatic_mutation.open.BRCA-EU.sorted.unique.bed results_all.formatted.txt /mnt/work1/users/lupiengroup/hawleyj/Data/wgEncodeUwDnaseMcf7HotspotsRep1.broadPeak 500000 0.7
                </code></pre>
            </section>

            <section id="2016-10_c3d-recommendations">
                <h3>C3D Recommendations</h3>

                <p>C3D is a script in its early stages. It requires a lot of code cleanup, and could do with adhering to some standard practices. Here are some things to do in the future to make the script better:</p>
                <ul>
                    <li>No config file</li>
                    <ul>
                        <li>Making config files and running scripts can get messy, since it requires writing to a config file every time you want to run the script</li>
                        <li>Using command line args is likely a better way to implement this, especially for making a pipeline</li>
                    </ul>
                    <li>Using getopt/standard argparsing</li>
                    <ul>
                        <li>Will make it easier to parse arguments without the need for a config file</li>
                        <li>Lessen chance of making mistakes when running</li>
                    </ul>
                    <li>Rewrite in Python or Perl</li>
                    <ul>
                        <li>Bash is useful, but not as readable or supported for many of these things</li>
                        <li>MuSE already uses perl, so it would be easier to make this an entire library</li>
                        <li>Calculations, command line, and parallelization are easily supported in these languages</li>
                    </ul>
                    <li>Parallelization</li>
                    <ul>
                        <li>These scripts take a long time to run</li>
                        <li>Using perl or python packages to parallelize computations will significantly speed up processing time</li>
                    </ul>
                    <li>Check files and folders before executing</li>
                    <ul>
                        <li>Nothing sucks more than having your script run for a few hours just to have it crash because you forgot to check that a directory or file existed</li>
                    </ul>
                </ul>
            </section>

        </section>
    </article>

    <footer>
        <a class="cd-top"><i class="glyphicon glyphicon-chevron-up"></i></a>
        <a class="cd-bottom"><i class="glyphicon glyphicon-chevron-down"></i></a>
    </footer>
</body>

</html>