#!/bin/bash
DIR=/mnt/work1/users/lupiengroup/TahmidProject/C3D/
#$ -cwd
#$ -N C3D
#$ -o logs/c3d.log
#$ -q lupiengroup
#$ -S /bin/bash

# Cross Cell-type Correlation in DNaseI hypersensitivity (C3D)
# Written by: Tahmid Mehdi
# Princess Margaret Cancer Centre - University Health Network, July 9, 2016

# Cross Cell-type Correlation in DNaseI hypersensitivity (C3D)
# Takes correlations between open regions of chromatin based on DNaseI hypersensitivity signals
# Regions with high correlations are candidates for 3D interactions
# Performs association tests on each candidate & adjusts p-values
# Identifies transcription factor motifs which overlap open regions
# Produces interaction landscapes and motif tracks in PDF format

# Help text
if [ "$1" = "-help" ]; then
	echo "Usage: sh c3d <config file>"
	echo "See template.cfg for an example of a config file"
	echo "Parameters for config file"
    echo "reference (Optional if matrix provided)\n\tReference sample BED/BEDGRAPH"
    echo "anchor (Required)\n\tFile with anchor regions BED (chr,start,stop,gene)"
    echo "db (Optional if matrix provided)\n\tFile containing list of background files BEDGRAPH"
    echo "outDirectory (Required)\n\tOutput directory"
    echo "matrix (Optional if reference and db provided)\n\tSignal data with regions as rows & samples as columns. It must have row names in chr:start-end format. If not provided,
 a matrix will be generated using reference & db"
    echo "window (Optional)\n\tNumber of flanking bps from anchor to check for open regions. If set to genome, it will take the correlation between each anchor DHS & every DHS in the reference. Defaults to 500000"
    echo "correlationThreshold (Optional)\n\tCorrelation threshold. Only interaction candidates with correlations above or equal to this will be shown. Defaults to 0"
    echo "pValueThreshold (Optional)\n\tP-value threshold. Only interaction candidates with p-values below or equal to this will be shown. Defaults to 1"
    echo "qValueThreshold (Optional)\n\tQ-value threshold. Only interaction candidates with q-values below or equal to this will be shown. Defaults to 1"
    echo "correlationMethod (Optional)\n\tCorrelation coefficient (pearson, spearman, kendall). Defaults to pearson"
	echo "motifs (Optional)\n\tFile containing list of JASPAR IDs of transcription factors (TFs) of interest. JASPAR_IDs has a key that shows which TFs each ID maps to"
	echo "fimoPValue (Optional)\n\tP-value threshold for motif occurances generated by FIMO for the TFs of interest. Defaults to 0.0001"
	echo "fimoQValue (Optional)\n\tQ-value threshold for motif occurances generated by FIMO for the TFs of interest. Defaults to 1"
	echo "figures (Optional)\n\ty if figures should be generated. Leave this parameter out if you do not want figures"
	echo "figureWidth (Optional)\n\tNumber of flanking bps from anchor to display on figures. Use a comma-separated list to assign different widths to different figures. For example, if you have 3 figures and 500000,100000,400 is passed then the figures will show 500000, 100000, and 400 bps around the first, second, and third anchors respectively. Defaults to window"
	echo "zoom (Optional)\n\tNumber of flanking bps from anchor to show for zoomed figures. You can also pass a comma-separated list of numbers if you want different zoom regions for each figure. For example, if you have 3 figures and you pass 10000,50000,75000 then the zoom regions for figures 1,2, and 3 will be 10000,50000, and 75000 respectively. Use 0 if you do not wish to zoom. Defaults to no zoom"
	echo "colours (Optional)\n\t4 colours for the interaction plots. For example, if blue,red,green,purple is passed then arcs of interactions with q-values>0.05 will be blue, between 0.05 and 0.01 will be red, 0.01 and 0.001 will be green and below 0.001 will be purple. Hexadecimal digits accepted. Defaults to shades of blue"
	echo "tracks (Optional)\n\ty if a file with tracks should be generated. This file can be uploaded to the UCSC Genome Browser to view interaction candidates. It creates the file by default. Change this parameter to anything but y if you do not want the file"
	echo "Multi-sample Parameters:"
	echo "Do not include reference or matrix if you want to run C3D on multiple samples"
	echo "references (Optional if matrices provided)\n\tA file containing a list of reference sample BED/BEDGRAPHs where each line is of the form '<reference.bed> <sample name>'. Each sample will be processed on a different node"
	echo "matrices (Optional if references and db provided)\n\tA file containing a list of reference sample matrices where each line is of the form '<signalMatrix.txt> <sample name>'. Each sample will be processed on a different node"
	echo "All other parameters are available"

	exit 1
fi 

function join_by { local IFS="$1"; shift; echo "$*"; }

# an array for parameters
typeset -A config
# default values
config=(
    	[reference]=""
    	[db]=""
    	[anchor]=""
	[outDirectory]=""
	[matrix]=""
	[motifs]=""
	[references]=""
	[matrices]=""
	[tracks]="y"
)
# read parameters from config file
while read line
do
	if echo $line | grep -F = &>/dev/null
    	then
		eval expanded_line="$line"
        	varname=$(echo "$line" | cut -d '=' -f 1)
        	config[$varname]=$(echo $expanded_line | cut -d '=' -f 2-)
    	fi
	if echo $line | grep -F 'module load' &>/dev/null
    	then
    		eval $line
	fi
done < $1

# if anchor or outDirectory are missing, warn user and stop script
if [ -z "${config[anchor]}" ] || [ -z "${config[outDirectory]}" ]; then
	echo "Missing anchor or outDirectory. Check config file"
	exit 1
fi
# if matrices is included
if [ "${config[matrices]}" != "" ]; then
	# arrays of files and sample names, respectively
        files=( $(cut -d ' ' -f1 ${config[matrices]} ) )
	sample=( $(cut -d ' ' -f2 ${config[matrices]} ) )
	# run C3D on a different node for each sample
        for i in "${!files[@]}"; do
                qsub $DIR/c3d.sh "$1" -matrix "${files[$i]}" -out "${config[outDirectory]}/${sample[$i]}" -sample "${sample[$i]}" -track "$((i + 1))" -numSamples "${#files[@]}" >> ${config[outDirectory]}/job_c3d.txt
        done
	jobID=( $(cut -d ' ' -f 3 ${config[outDirectory]}/job_c3d.txt) )
	if [ "${config[tracks]}" = "y" ]; then
		# make a custom track that combines samples & adds motif occurances
        	qsub -hold_jid $(join_by , "${jobID[@]}") -cwd $DIR/makeTracks.sh "${config[outDirectory]}/anchors.bed" "${config[outDirectory]}" "${config[matrices]}" "${config[motifs]}"
	fi
# if references is included
elif [ "${config[references]}" != "" ]; then
	# if no db passed, stop script & warn user
        if [ -z "${config[db]}" ]; then
                echo "Missing db. Check config file. Add db or matrices"
                exit 1
	else
		files=( $(cut -d$'\t' -f1 ${config[references]} ) )
	        sample=( $(cut -d$'\t' -f2 ${config[references]} ) )

        	for i in "${!files[@]}"; do
                	qsub $DIR/c3d.sh "$1" -ref "${files[$i]}" -out "${config[outDirectory]}/${sample[$i]}" -sample "${sample[$i]}" -track "$((i + 1))" -numSamples "${#files[@]}" >> ${config[outDirectory]}/job_c3d.txt
		done
		jobID=( $(cut -d ' ' -f 3 ${config[outDirectory]}/job_c3d.txt) )
        	if [ "${config[tracks]}" = "y" ]; then
                	qsub -hold_jid $(join_by , "${jobID[@]}") -cwd $DIR/makeTracks.sh "${config[outDirectory]}/anchors.bed" "${config[outDirectory]}" "${config[references]}" "${config[motifs]}"
        	fi
	fi
else
	# if matrix is missing and reference and db are not supplied, warn user and stop script
	if [ -z "${config[matrix]}" ]; then
        	if [ -z "${config[reference]}" ] || [ -z "${config[db]}" ]; then
                	echo "Missing reference or db. Check config file. Add reference and db or matrix"
                	exit 1
        	fi
	fi
	# run C3D without overwrite parameters because there's only 1 sample
	qsub $DIR/c3d.sh "$1" > ${config[outDirectory]}/job_c3d.txt
	jobID=$(cut -d ' ' -f 3 ${config[outDirectory]}/job_c3d.txt)

	if [ "${config[tracks]}" = "y" ]; then
                qsub -hold_jid $jobID -cwd $DIR/makeTracks.sh "${config[outDirectory]}/anchors.bed" "${config[outDirectory]}" "${config[references]}" "${config[motifs]}"
        fi
fi

